pkg/reconciler/route/traffic/rollout.go

```diff
@@ -70,3 +75,129 @@ type RevisionRollout struct {
 	// target percentage.
 	Percent int `json:"percent"`
 }
+
+// Step merges this rollout object with the  previous state and
+// returns a new Rollout object representing the merged state.
+// At the end of the call the returned object will contain the
+// desired traffic shape.
+// Step will return cur if no previous state was available.
+func (cur *Rollout) Step(prev *Rollout) *Rollout {
+	if prev == nil || len(prev.Configurations) == 0 {
+		return cur
+	}
+
+	// The algorithm below is simplest, but probably not the most performant.
+	// TODO: optimize in the later passes.
+
+	// Map the configs by tag.
+	currConfigs, prevConfigs := map[string][]*ConfigurationRollout{}, map[string][]*ConfigurationRollout{}
+	for i, cfg := range cur.Configurations {
+		currConfigs[cfg.Tag] = append(currConfigs[cfg.Tag], &cur.Configurations[i])
+	}
+	for i, cfg := range prev.Configurations {
+		prevConfigs[cfg.Tag] = append(prevConfigs[cfg.Tag], &prev.Configurations[i])
+	}
+
+	var ret []ConfigurationRollout
+	for t, ccfgs := range currConfigs {
+		pcfgs, ok := prevConfigs[t]
+		// A new tag was added, so we have no previous state to roll from,
+		// thus just add it to the return, we'll rollout to 100% from the get go (and it is
+		// always 100%, since default tag is _always_ there).
+		// So just append to the return list.
+		if !ok {
+			ret = append(ret, *ccfgs[0])
+			continue
+		}
+		// This is basically an intersect algorithm,
+		// It relies on the fact that inputs are sorted.
+		i := 0
+		for j := 0; i < len(ccfgs) && j < len(pcfgs); {
+			switch {
+			case ccfgs[i].ConfigurationName == pcfgs[j].ConfigurationName:
+				// Config might have 0% traffic assigned, if it is a tag only route (i.e.
+				// receives no traffic via default tag).
+				if ccfgs[i].Percent != 0 {
+					ret = append(ret, *stepConfig(ccfgs[i], pcfgs[j]))
+				}
+				i++
+				j++
+			case ccfgs[i].ConfigurationName < pcfgs[j].ConfigurationName:
+				// A new config, has been added. No action for rollout though.
+				ret = append(ret, *ccfgs[i])
+				i++
+			default: // cur > prev.
+				// A config has been removed during this update.
+				// Again, no action for rollout, since this will no longer
+				// be rolling it out (or sending traffic to it overall).
+				j++
+			}
+		}
+		// Keep the remaining new objects
+		for ; i < len(ccfgs); i++ {
+			ret = append(ret, *ccfgs[i])
+		}
```

Comment:
```suggestion
		for i, j := 0, 0; i < len(ccfgs); {
			switch {
			case j == len(pcfgs):
				// TODO(vagababov): useful comment
				ret = append(ret, *ccfgs[i])
				i++
			case ccfgs[i].ConfigurationName == pcfgs[j].ConfigurationName:
				// Config might have 0% traffic assigned, if it is a tag only route (i.e.
				// receives no traffic via default tag).
				if ccfgs[i].Percent != 0 {
					ret = append(ret, *stepConfig(ccfgs[i], pcfgs[j]))
				}
				i++
				j++
			case ccfgs[i].ConfigurationName < pcfgs[j].ConfigurationName:
				// A new config, has been added. No action for rollout though.
				ret = append(ret, *ccfgs[i])
				i++
			default: // cur > prev.
				// A config has been removed during this update.
				// Again, no action for rollout, since this will no longer
				// be rolling it out (or sending traffic to it overall).
				j++
			}
		}
```

Reply:
hmmm, not sure if this reads better.

Reply:
Anyway done.